### ISSUE TITLE
wechat6.5？

### ISSUE THREAD

**Issue by pxss** [ppwwyyxx/wechat-dump#42]
wechat6.5 how to decrypt  ？

**Comment by rom1504** [comment]
still the same

**Comment by ChiChou** [comment]
WeChat 6.5 generates a random string as device id. 

Decompiled code in this class `com.tencent.mm.compatible.d.p`:

```Java
    private static String rZ() {
        String v0;
        long v8 = 13802011623424L;
        int v7 = 102833;
        int v6 = 15;
        GMTrace.i(v8, v7);
        String v1 = bf.ec(com.tencent.mm.sdk.platformtools.aa.getContext());
        if(v1 == null || v1.length() <= 0) {
            Random v3 = new Random();
            v3.setSeed(System.currentTimeMillis());
            v0 = "A";
            int v1_1 = 0;
            while(v1_1 < v6) {
                ++v1_1;
                v0 = v0 + (((char)(v3.nextInt(25) + 65)));
            }
        }
        else {
            v0 = "A" + v1 + "123456789ABCDEF".substring(0, v6);
        }

        com.tencent.mm.sdk.platformtools.v.w("MicroMsg.DeviceInfo", "generated deviceId=" + v0);
        GMTrace.o(v8, v7);
        return v0;
    }
```

The string is stored in `/data/data/com.tencent.mm/MicroMsg/CompatibleInfo.cfg`, which is a  serialized Java HashMap. Simply use strings to extract the plain text (may not work):

```
shamu:/data/data/com.tencent.mm # strings ./MicroMsg/CompatibleInfo.cfg | grep -oE '^A[[:upper:]]+'"

```

Or, use a instrument tool like [frida](https://www.frida.re) then access its method with Reflect:

```
$ frida -U com.tencent.mm

[USB::Motorola Nexus 6::com.tencent.mm]-> Java.perform(function() {
                                            var k = Java.use('com.tencent.mm.compatible.d
.k');
                                            console.log(k.rE().get(256));
                                          });
```

**Comment by ppwwyyxx** [comment]
@ChiChou Thanks very much for the details!! I'll try your method and update the code.

**Comment by ChiChou** [comment]
Seems like password generation is more complicated now. Will post the whole decryption loop when I'm done.

**Comment by ChiChou** [comment]
@ppwwyyxx Damn I was fooled. WeChat 6.5 uses the same password generator. 

![ced409b5-762a-4b5b-b397-3809604d1937](https://cloud.githubusercontent.com/assets/2802045/25553252/1d24a6f8-2ce1-11e7-95b4-e4e414cb6bb2.png)


Sorry for misleading.

**Comment by ppwwyyxx** [comment]
So I only need to change the device id (previously IMEI number), right?

**Comment by ChiChou** [comment]
This may work:
https://gist.github.com/ChiChou/36556fd412a9e3216abecf06e084e4d9

It invokes JNI to unserialize the `MicroMsg/CompatibleInfo.cfg` file, which stores the actual device id (IMEI or whatever it generates).

**Comment by ppwwyyxx** [comment]
Decryption seems to work for me after I tried other methods to get IMEI.

**Comment by tony-zdl** [comment]
@ppwwyyxx which method did you try to get IMEI?

**Comment by ppwwyyxx** [comment]
In the README.

**Comment by 8enet** [comment]
@ChiChou Thanks, it work fine for me!
get the IMEI unexpectedly is `1234567890ABCDEF`. :joy:

**Comment by cs-usst** [comment]
@8enet How does that file work? I don't understand.

**Comment by ChiChou** [comment]
@cs-usst 

The device id that WeChat uses does not always match IMEI. But WeChat will always store this id in file `/data/data/com.tencent.mm/MicroMsg/CompatibleInfo.cfg`, which contains a `HashMap<Integer, Object>` serialized with `ObjectOutputStream`. I wrote the gist with C to call Java runtime (`void device_id(char *id, size_t len)`).

Anyways, Java serialized string still looks as is, so simply pull this file from device and `strings CompatibleInfo.cfg` to pick it.

**Comment by ChiChou** [comment]
Tried deserialize that with python, but the `javaobj` from pypi looks buggy

**Comment by prusswan** [comment]
@QiangF assuming your dump is correct, the IMEI should be AXKXESIIOBTUPLVJ (the sq bit is an artifact from improper deserialization - a proper way to do it can be found here: https://blog.csdn.net/jiezhi2013/article/details/22181771)

**Comment by amakuramio** [comment]
How to run on windows?

**Comment by ChiChou** [comment]
> How to run on windows?

https://docs.microsoft.com/en-us/windows/wsl/wsl2-install

**Comment by amakuramio** [comment]
Help me please
I tried to compile wechat xposed module to fetch pass but didnt work
Im stupid, I have a very old laptop so slow
I used  imei of my first and second phone, fail
I cant decrypt!
https://drive.google.com/file/d/1fNYFBsIxBgCXZ4cywUe26KpSJeBHXxjY/view?usp=drivesdk
Please anyone can help me decrypt mine and send me the contents!

**Comment by amakuramio** [comment]
> @QiangF assuming your dump is correct, the IMEI should be AXKXESIIOBTUPLVJ (the sq bit is an artifact from improper deserialization - a proper way to do it can be found here: https://blog.csdn.net/jiezhi2013/article/details/22181771)

Tried to run in windows, said that maptest is missing

**Comment by amakuramio** [comment]
> Decryption works for me after I tried other methods to get IMEI.

i installed ubuntu, i also have compatibleinfo, now what do i do
which commands? what do i download?

**Comment by amakuramio** [comment]
> This may work:
> https://gist.github.com/ChiChou/36556fd412a9e3216abecf06e084e4d9
> 
> It invokes JNI to unserialize the `MicroMsg/CompatibleInfo.cfg` file, which stores the actual device id (IMEI or whatever it generates).

how do i use this one? add to the folder wechat_dump then strings .. cfg?

**Comment by amakuramio** [comment]
ridiculous, it was the original phone, i registered with in 2016, plus uim, and the 7 md5 thingy
but i always used capital letters .................................................
i got it, decrypted... haha im so stupid

**Comment by altayhunter** [comment]
When I deserialize the HashMap in my CompatibleInfo.cfg I get the following: `{98305=false, 94209=0}`  
The bytes between the instances of sq are `00 7E 00 02 00 01 70 01`. Any ideas?

**Comment by ppwwyyxx** [comment]
The latest decryption script (https://github.com/ppwwyyxx/wechat-dump/blob/b2468073045cc256f6f57cb7614eb66efef9c923/decrypt-db.py#L1) now implements the logic to obtain device id from CompatibleInfo.cfg automatically.

**Comment by wzyboy** [comment]
Is anyone using Android 10? In Android 10, the IMEI is not available to third-party apps.

I tried to extract whatever WeChat gets by deserializing `/data/data/com.tencent.mm/MicroMsg/CompatibleInfo.cfg` file but it turned out that this file is empty (file size 0 byte).

**Comment by Admingyu** [comment]
> Is anyone using Android 10? In Android 10, the IMEI is not available to third-party apps.
> 
> I tried to extract whatever WeChat gets by deserializing `/data/data/com.tencent.mm/MicroMsg/CompatibleInfo.cfg` file but it turned out that this file is empty (file size 0 byte).

Me too, I got the imei in `com.tencent.mm/shared_prefs/WLOGIN_DEVICE_INFO.xml`：

```xml
<?xml version='1.0' encoding='utf-8' standalone='yes' ?>
<map>
    <string name="imei">fb8da1fe1cb980202656xxxxxxxx</string>
</map>

```

But still not working

**Comment by ppwwyyxx** [comment]
The above comment was already addressed in https://github.com/ppwwyyxx/wechat-dump/issues/70

**Comment by Admingyu** [comment]
> The above comment was already addressed in #70

Thanks a lot !!!
